### ISSUE TITLE
Tracking issue for `#![feature(const_precise_live_drops)]`

### ISSUE THREAD

**Issue by ecstatic-morse** [rust-lang/rust#73255]
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/issues/73255)*

This is a tracking issue for a more precise version of checking for drops in `const` contexts.

The feature gate for the issue is `#![feature(const_precise_live_drops)]`.

With this feature enabled, drops are checked on slightly elaborated MIR. This makes the analysis that prevents dropping in `const fn` more accurate for code that does not actually drop anything: specifically, if the initial MIR contains an unnecessary call to `drop` that may be eliminated in elaboration, it can be accepted with `const_precise_live_drops`.

Internally, the library can also use this on a per-function basis with `#[rustc_allow_const_fn_unstable(const_precise_live_drops)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.
Discussion comments will get marked as off-topic or deleted.
Repeated discussions on the tracking issue may lead to the tracking issue getting locked.

### Steps

- [x] implementation:  #71824

### Blockers and Concerns:

- https://github.com/rust-lang/rust/issues/91009
- Changes to what exact elaboration occurs may change what we accept
- (per Ralf on Zulip): ideally this would run on the same MIR and with the same logic as borrowck, but that is hard.

**Comment by RalfJung** [comment]
Cc @rust-lang/wg-const-eval (which I think never happened for this feature or the PR)

**Comment by DoumanAsh** [comment]
Is there particular reason why this was moved to be a feature?
It is more or less bug fix to inaccurate drop detection.
So shouldn't it be already stable?
This would make it easier to create builders with generic parameters as we currently cannot mutate `self`

**Comment by oli-obk** [comment]
cc @rust-lang/lang I am nominating this feature for stabilization

citing @ecstatic-morse from the impl PR:

This isn't really a "feature" but a refinement to const-checking, and it's kind of non-obvious what the user opts in to with the feature gate, since drop elaboration is an implementation detail. A relevant precedent is `#![feature(nll)]`, but that was much larger and more fundamental than this change would be.

This is also somewhat relevant to the stabilization of `#![feature(const_if_match)]`. Nothing in this PR is specifically related to branching in constants. For instance, it makes the example below compile as well. However, I expect users will commonly want to move out of `Option::<T>::Some` within a `match`, which won't work without this PR if `T: !Copy`

```rust
const _: Vec<i32> = {
    let vec_tuple = (Vec::new(),);
    vec_tuple.0
};
```

> To clarify, this PR makes that code compile because previously we were dropping the vec_tuple which had type Option<(Vec<i32>,)>?

This code is functionally no different from the following, which currently works on stable because `x` is moved into the return place unconditionally.

```rust
const X: Option<Vec<i32>> = { let x = Vec::new(); x };
```

Const-checking only considers whole locals for const qualification, but drop elaboration is more precise: It works on projections as well. Because drop elaboration sees that all fields of the tuple have been moved from by the end of the initializer, no `Drop` terminator is left in the MIR, despite the fact that the tuple *itself* is never moved from.

**Comment by RalfJung** [comment]
@oli-obk how do things look like in implementation complexity and the risk of locking us into a scheme that might be hard to maintain or make compatible with other future extensions?

**Comment by DoumanAsh** [comment]
I believe implementation complexity should not be an issue considering current behavior is bug as compiler incorrectly reports error on drop being executed in const fn.
Such things ought to be fixed in any case. Unless we're fine with compiler  incorrectly reporting error as right now?

**Comment by oli-obk** [comment]
> I believe implementation complexity should not be an issue considering current behavior is bug as compiler incorrectly reports error on drop being executed in const fn.
> Such things ought to be fixed in any case. Unless we're fine with compiler incorrectly reporting error as right now?

We may be able to achieve the same result (allowing these kind of `const fn`) with a different scheme. So Ralf's question is absolutely reasonable to ask and I will answer it after reviewing the implementation again in detail so I can give an educated answer.

**Comment by RalfJung** [comment]
@DoumanAsh due to the halting problem, rustc will *always* report "incorrect errors". It is impossible to predict at compiletime if e.g. `drop` will be executed when a piece of code is being run. Similarly, the borrow checker will always reject correct programs. That's just a fact of programming languages.

So there's always a trade-off between analysis complexity and "incorrect errors". My impression after looking at the PR here is that the analysis complexity is totally reasonable (it mostly implicitly reuses the analysis performed by drop elaboration), but this is still a decision we should make explicitly, not implicitly.

@oli-obk Thanks. :) As you probably saw, I also [left some questions in the PR that implemented the analysis](https://github.com/rust-lang/rust/pull/71824#pullrequestreview-611608120).

**Comment by oli-obk** [comment]
I did see these comments, just didn't have time to look more deeply yet.

Cross posting from the comments on that PR:

I am guessing the difference between the `needs_drop` analysis, and drop elaboration's use of `MaybeInitializedPlaces` and `MaybeUninitializedPlaces` is the reason that this feature gate exists at all. We could probably rewrite drop elaboration in terms of the `NeedsDrop` qualif, which would (afaict) allow `post_drop_elaboration` to not do any checks except for looking for `Drop` terminators.

Of course such a change would insta-stabilize the feature gate from this PR without any reasonable way to avoid said stabilization. So I'm tempted to stabilize the feature but leave a FIXME on this function to merge its qualif checks into `elaborate_drops`

**Comment by RalfJung** [comment]
> We could probably rewrite drop elaboration in terms of the NeedsDrop qualif, which would (afaict) allow post_drop_elaboration to not do any checks except for looking for Drop terminators.
> Of course such a change would insta-stabilize the feature gate from this PR without any reasonable way to avoid said stabilization. 

Even worse, it would change drop elaboration, which is a rather tricky part of the compiler.^^ That should be done with utmost care.

**Comment by Mark-Simulacrum** [comment]
Dropping the nomination here as it's waiting on Niko; that's tracked in the language team action item list.

**Comment by inquisitivecrystal** [comment]
What remains to be done before this can be stabilized?

**Comment by ecstatic-morse** [comment]
Hey y'all. I'll try to summarize exactly what this feature gate does to help inform a decision on stabilization. Oli and Ralf probably already know this information, and can skip the next three paragraphs.

Currently, it is forbidden to run custom `Drop` implementations during const-eval. This is because `Drop::drop` is a trait method, and there's no (stable) way to declare that trait methods are const-safe. MIR uses a special terminator (`Drop`) to indicate that something is dropped. When building the MIR initially, a `Drop` terminator is inserted for every variable that has drop glue (`ty.needs_drop()`). Typically, const-checks are run on the newly built MIR. This is to prevent MIR optimizations from eliminating or transforming code that would fail const-checking into code that passes it, and would make compilation success dependent on MIR optimizations, which we try very hard to avoid.

However, some `Drop` terminators that we generate during MIR building will never actually execute. For example,
  - A `struct` has a field with a custom `Drop` impl (e.g. `Foo(Box<i32>)`), but that field is always moved out of before the variable goes out of scope.
  - An `enum` has a variant with a custom `Drop` impl, but we are in a `match` arm for a different variant without a custom `Drop` impl.

We rely on a separate MIR pass, `elaborate_drops`, to remove these terminators prior to codegen (it also does some extra stuff, like adding dynamic drop flags). This happens pretty early in the optimization pipeline, but it is still an optimization, so it runs after const-checking. As a result, the stable version of const-checking sees some `Drop` terminators that will never actually run, and raises an error.

`#![feature(const_precise_live_drops)]` defers the `Drop` terminator const-checks until *after* drop elaboration is run. As I mention above, this is unusual, since it makes compilation success dependent on the result of an optimization. On the other hand,
  - Drop elaboration doesn't change very often. Unlike some other optimizations, it is conceptually pretty simple to determine whether a `Drop` terminator is frivolous along a given code path. It depends solely on whether some set of move paths are initialized at that point in the CFG.
  - It is always profitable to remove drop terminators from the MIR when we can. This makes it unlikely that we will ever want to "dial back" drop-elaboration.
   - Despite being conceptually straightforward, drop elaboration is not free to compute and having two separate implementations of it (one in const-checking and one in the optimization pipeline) is both inefficient and bad for maintainability.
   - If there *is* bug in drop elaboration that causes us to wrongly eliminate `Drop`s, wrongly accepting some `const fn`s will be the least of our worries, relatively speaking.

For these reasons, I chose the current implementation strategy, which allows (among others) `Option::unwrap` to be `const fn`. I realize that none of these arguments are dispositive, and I don't think it's unreasonable to gate stabilization of this feature on reimplementing the relevant bits of drop elaboration inside const-checking, although I still think it's overly conservative.

Besides that big question, I think there were also some concerns from `const-eval` members around the use of the `NeedsDrop` qualif and test coverage (e.g. https://github.com/rust-lang/rust/pull/71824#discussion_r594392980). I'll try to answer those so they can provide a recommendation.

**Comment by inquisitivecrystal** [comment]
I'm not educated on the details, but it would be super nice to see this stabilized in some form. There are a comparatively large number of new APIs that rely on this. For examples, see issues #76654, #82814, and PR #84087, the last of which is an approved stabilization PR that can't be merged until this is stabilized. 

That's why I was checking in on the progress towards stabilization a few days ago. I'm sorry about that, by the way. I know that that sort of message can be annoying, but I wanted to know there if there was anything I could do to help move this along.

**Comment by ecstatic-morse** [comment]
The fact that the main blocker (and why this was feature gated in the first place) is the implementation makes this somewhat unusual (see https://github.com/rust-lang/rust/pull/71824#discussion_r421675954). That makes it more the domain of the compiler-team rather than the lang-team. Niko reviewed #71824 and is assigned to this issue, but I'm hesitant to ping them specifically unless their expertise is required.

So, if you want to see this stabilized I would figure some process for getting consent from the compiler team (Oli is a member, so there's at least one). They might require  documenting the drop-elaboration/const-checking dependency in the dev-guide and maybe the module itself, which I'm happy to do if asked. After that, I can write a stabilization report with some examples and we can do lang-team FCP (assuming any lingering concerns from @rust-lang/wg-const-eval have been addressed).

I'm, uh, not great at navigating bureaucratic systems with many veto points, so if you want to drive this forward your help would be greatly appreciated. However, unless we end up reimplementing drop-elaboration in const-checking like I mention above, I don't think much of the remaining work is technical.

**Comment by oli-obk** [comment]
I think your summary post contains most of what we need for a stabilization report (full instructions [here](https://rustc-dev-guide.rust-lang.org/stabilization_guide.html?highlight=stabilization%20report#write-a-stabilization-report)). We can just mark this issue as requiring sign-off from both T-compiler and T-lang and do all of this at once.

**Comment by RalfJung** [comment]
> because the MIR still contains a Drop terminator on that arm

Specifically, the "early" MIR that most of const checking runs on still has it. After drop elaboration, it is gone, which is why running the 'drop' part of const checking later makes such a difference.

**Comment by ecstatic-morse** [comment]
True. I've always been unclear about how much work is duplicated between drop elaboration and the borrow checker, since the latter runs first. Perhaps Niko or Felix could say more?

**Comment by nagisa** [comment]
From what I remember borrowck relies upon the variable liveness annotations (`StorageLive`/`StorageDead`) and does not rely on drop elaboration in any way or form.

**Comment by nikomatsakis** [comment]
@rfcbot concern reference-material

I'm raising @cramertj's point to an actual concern. I don't think we need a "formal specification" but I do think we need documentation in the reference. It doesn't have to be enough to implement things independently, but I think that it should be enough that humans can understand it. @ecstatic-morse let's sync up about it a bit.

**Comment by ecstatic-morse** [comment]
I opened rust-lang/rustc-dev-guide#1240 last week.

To add some more context to [this comment](https://github.com/rust-lang/rust/issues/73255#issuecomment-943840387) for  @nikomatsakis (in case they declared notification bankruptcy). Like the const-checker, the borrow-checker also needs to ignore drops of places that are known to be uninitialized, and it also runs before drop elaboration. I believe it runs `MaybeUninitializedPlaces` solely to remove dead drop terminators. This is a bit less onerous for the borrow-checker than in the const-checker, since the former already has the `MoveData` (which contains the move path tree, etc).

To me, this suggests that we should do dead drop removal as a pre-pass for both the borrow-checker and the const-checker but keep full drop elaboration as an optimization. However, I remember false edges making things more complex: That's why borrow-checker does not use [`mark_inactive_variants_as_uninit`](https://github.com/rust-lang/rust/blob/4961b107f204e15b26961eab0685df6be3ab03c6/compiler/rustc_mir_dataflow/src/impls/mod.rs#L131-L140), for example. Do you think making "remove dead drops" a separate pass is feasible?

**Comment by RalfJung** [comment]
> Do you think making "remove dead drops" a separate pass is feasible?

Or, alternatively, it might be nice to be able to share the results of `MaybeUninitializedPlaces` between borrowck and constck. If they are computed anyway, we might as well use them twice...

**Comment by nbdd0121** [comment]
I am not sure why this code is accepted:
```rust
#![feature(const_precise_live_drops)]

struct S;

impl Drop for S {
    fn drop(&mut self) {
        println!("Hello!");
    }
}

const fn foo() {
    let s = S;
}
```

When trying to use it in const context it just says "error: any use of this value will cause an error" without any other errors generated.

**Comment by ecstatic-morse** [comment]
@nbdd0121 Could you open a new issue? The final "post-borrowck cleanup" pass is the [Deaggregator](https://github.com/rust-lang/rust/blob/0701419e96d94e5493c7ebfcecb66511ab0aa778/src/librustc_mir/transform/deaggregator.rs), which replaces the assignment `_1 = S;` with a nop because `S` has no fields. That means S` is never initialized in the MIR before its `Drop` terminator, so it never gets `NeedsDrop`.

**Comment by ecstatic-morse** [comment]
In addition to the bug found by Gary above, we also had #90752 filed yesterday, a bug in drop elaboration proper. Time to test this part out:

> If there *is* bug in drop elaboration that causes us to wrongly eliminate `Drop`s, wrongly accepting some `const fn`s will be the least of our worries, relatively speaking.

I recommend we don't stabilize `#![feature(const_precise_live_drops)]` as-is but instead pursue something like what I mentioned [here](https://github.com/rust-lang/rust/issues/73255#issuecomment-961354314):

>  this suggests that we should do dead drop removal as a pre-pass for both the borrow-checker and the const-checker but keep full drop elaboration as an optimization

**Comment by nikomatsakis** [comment]
Based on the previous comment, cancelling stabilization proposal:

@rfcbot fcp cancel

**Comment by RalfJung** [comment]
This is probably also a blocker: https://github.com/rust-lang/rust/issues/91009

**Comment by scottmcm** [comment]
I noticed in [a PR today](https://github.com/rust-lang/rust/pull/91222/files#diff-1a068eac2c0aea102f63d52824630fcba21adc6431426b70c9769137e2ff42a7R11) that I accidentally changed what this was doing by making what I *thought* was just an optimization -- same as `SimplifyBranches` -- but changed the precise drop errors.  So I'm glad to see the cancel here, since it made me nervous about where this was happening in the pipe.

**Comment by RalfJung** [comment]
I wonder if it is possible to make our existing const drop checks a bit smarter so that at least `Option::unwrap` can be accepted, without the future-compat hazard of letting dropck do the checking? It would be great to unblock (parts of) https://github.com/rust-lang/rust/issues/67441.

**Comment by clarfonthey** [comment]
Is this still blocked? It appears that #91009 has been closed.

The ability to make `unwrap` work in `const fn` would help a lot with simplifying a lot of existing code, since it can help replace a lot of manual match statements with `unwrap`s.

**Comment by RalfJung** [comment]
We still have this problem: https://github.com/rust-lang/rust/issues/73255#issuecomment-979808791. Basically exactly the thing we were worried about (depending on subtle dropck details) actually came up just after the attempt to stabilize.

**Comment by clarfonthey** [comment]
Ah! Is there a dedicated issue open for that that can be linked in the description, or is that just a known issue at the moment?

**Comment by RalfJung** [comment]
It's not a very concrete issue, and I don't think is tracked anywhere explicitly outside of this tracking issue.

**Comment by clarfonthey** [comment]
That's fair. As is expected with all these const features, something subtle and complicated lurks in the depths that makes it hard to finish up.

I was kinda hopeful that this was mostly done, but alas.

**Comment by RalfJung** [comment]
It might be, I am honestly not familiar enough with drop elaboration to really evaluate the trade-offs here.

I hope someone else reading along has some good ideas for what can be done before we ask the lang team to discuss this again.

**Comment by RalfJung** [comment]
One thing I did miss when writing the comments above is https://github.com/rust-lang/rust/pull/91410: this feature no longer relies on full drop elaboration, just a lightweight version of dead drop removal.

Still, doing this somewhere in the middle of our MIR pipeline does not feel great. Instead maybe it is possible to run the analysis that remove_uninit_drops performs during const-checking so that we can just ask "can this drop ever happen" rather than manifesting the result as explicit MIR? Then this could just happen during regular const checking.

(Borrowck already does something like that, doesn't it? If someone knows a borrowck expert, please point them to this thread, maybe they can help :)

**Comment by nikomatsakis** [comment]
These days I've been thinking that we should reformulate borrow check (and drop check) to take place on THIR. I plan to prototype this in a-mir-formality, which I suppose may have to be renamed to just formality as a result... ;'(. Seems relevant.
